dev/com.ibm.ws.security.authentication/src/com/ibm/ws/security/authentication/cache/CacheObject.java

```diff
@@ -59,7 +69,62 @@ public void addLookupKey(Object key) {
         return lookupKeys;
     }
 
+    /**
+     * Get the {@link Subject} stored in this {@link CacheObject}.
+     *
+     * @return The subject.
+     */
     public Subject getSubject() {
         return this.subject;
     }
+
+    @SuppressWarnings("unchecked")
+    private void readObject(ObjectInputStream input) throws ClassNotFoundException, IOException {
+        /*
+         * Read all non-transient fields.
+         */
+        input.defaultReadObject();
+
+        /*
+         * Read the fields comprising the Subject.
+         */
+        boolean subjectReadOnly = input.readBoolean();
+        Set<Principal> subjectPrincipals = (Set<Principal>) input.readObject();
+        Set<Object> subjectPubCreds = (Set<Object>) input.readObject();
+        Set<Object> subjectPrivCreds = (Set<Object>) input.readObject();
+
+        /*
+         * Create the subject from the subject fields.
+         */
+        subject = new Subject(subjectReadOnly, subjectPrincipals, subjectPubCreds, subjectPrivCreds);
+    }
+
+    private void writeObject(ObjectOutputStream output) throws IOException {
+        /*
+         * Write all non-transient fields.
+         */
+        output.defaultWriteObject();
+
+        /*
+         * Write the fields comprising the Subject.
+         */
+        output.writeBoolean(subject.isReadOnly());
+        output.writeObject(copySet(subject.getPrincipals()));
+        output.writeObject(copySet(subject.getPublicCredentials()));
+        output.writeObject(copySet(subject.getPrivateCredentials()));
+    }
+
+    /**
+     * Copy the contents of the input {@link Set} into a new {@link Set} instance. This is necessary for serializing the
+     * principals and credentials of the {@link Subject} since the SecureSet implementation was not intended for serialization.
+     *
+     * @param <T> The type for the set to copy.
+     * @param set The set to copy.
+     * @return The copied set.
+     */
+    private static <T> Set<T> copySet(Set<T> set) {
+        Set<T> copy = new HashSet<T>();
+        copy.addAll(set);
+        return copy;
```

Comment:
```suggestion
        return new HashSet<T>(set);
```
By calling this constructor it does the sizing better.

Reply:
Removed the copySet method and just called the constructor instead.